### PR TITLE
[QEff Finetune] : Made fixes to training script

### PR DESCRIPTION
**Pull request by quic-mamta**
- Added padding of dataset to make it divisible by batch_size * num_devices.
- Updated dataloader code to prevent dropping the last slice from the dataset.
- Updated samsum dataset name from "Samsung/samsum" to "knkarthick/samsum".
- Added loss modified loss functions to support the loss_weight parameter.
- Refactored and corrected the logging of loss and ppl across devices.
 
Below are the numbers with this PR:
 
Dataset: Samsum
Model: Llama-3.2-1B
Epoch: 1
 
| Sr. No. | # Devices | Grad. Accum. | BS | Global BS | Train Loss | Test Loss | Train PPL | Test PPL |
|--------:|----------:|-------------:|---:|-----------:|------------:|-----------:|-----------:|----------:|
|       1 |        48 |            4 |  1 |        192 |      1.3206 |     1.2855 |     3.7471 |    3.6360 |

Dataset: Samsum
Model: Llama-3.1-8B
Epoch: 1
 
| Sr. No. | # Devices | Grad. Accum. | BS | Global BS | Train Loss | Test Loss | Train PPL | Test PPL |
|--------:|----------:|-------------:|---:|-----------:|------------:|-----------:|-----------:|----------:|
|       1 |         4 |            4 |  1 |        16  |      1.0120 |     1.0725 |     2.7510 |    2.9234 |

### CODE REVIEW

File: QEfficient/finetune/utils/train_utils.py [hunk 1]
```diff
@@ -235,11 +241,23 @@ def train(
                 train_step_metric.append(step_metric_val)
 
             if train_config.grad_scaler:
-                scaler.scale(loss).backward()  # backward pass
+                if train_config.enable_ddp:
+                    with model.no_sync():
```
Comment (quic-swatia) [inline QEfficient/finetune/utils/train_utils.py:245]: This will result in no syncing of gradients at any step.

File: QEfficient/finetune/utils/train_utils.py [hunk 2]
```diff
@@ -235,11 +241,23 @@ def train(
...
+                        scaler.scale(loss).backward()  # backward pass
+                else:
+                    scaler.scale(loss).backward()  # backward pass
             else:
-                loss.backward()  # backward pass
+                if train_config.enable_ddp:
+                    # FIXME: We can not stop transfer of gradient across devices every time.
+                    # In grad accumulation last step should transfer gradients across devices.
+                    with model.no_sync():
```
Comment (quic-swatia) [inline QEfficient/finetune/utils/train_utils.py:253]: This will result in no syncing of gradients at any step here as well.

File: QEfficient/finetune/dataset/samsum_dataset.py [hunk 3]
```diff
@@ -6,10 +6,11 @@
 # -----------------------------------------------------------------------------
 
 import datasets
+from QEfficient.finetune.dataset.helper import IGNORE_INDEX
 
 
 def get_preprocessed_samsum(dataset_config, tokenizer, split, context_length=None):
-    dataset = datasets.load_dataset("Samsung/samsum", split=split, trust_remote_code=True)
+    dataset = datasets.load_dataset("knkarthick/samsum", split=split, trust_remote_code=True)
```
[the remark this answers is not in the document]
Comment (quic-meetkuma) [reply QEfficient/finetune/dataset/samsum_dataset.py:14]: We are not distributing this dataset hence, it should not be a problem.

File: QEfficient/finetune/utils/config_utils.py [hunk 4]
```diff
@@ -115,10 +115,26 @@ def generate_dataset_config(dataset_name: str) -> Any:
...
+        dataset = dataset.add_item(sample_input) 
+    return dataset  
+
+
 def get_dataloader_kwargs(train_config, dataset, dataset_processer, mode):
     kwargs = {}
     batch_size = train_config.batch_size_training if mode == "train" else train_config.val_batch_size
     if train_config.enable_ddp:
+        print("Length of dataset before: ", len(dataset))
+        dataset = pad_dataset(dataset, batch_size, 2)
```
[the remark this answers is not in the document]
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/config_utils.py]: Updated in latest.

File: QEfficient/finetune/utils/train_utils.py (hunk 2 above)
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/train_utils.py]: Yes, correct. Removed this no_sync change from this PR. We will raise separate PR for that.
